validation/src/test/scala/hmda/validation/rules/lar/macro/Q070Spec.scala

```diff
@@ -0,0 +1,132 @@
+import hmda.model.fi.SubmissionId
+import hmda.model.fi.lar.LoanApplicationRegister
+import hmda.model.institution.Institution
+import hmda.validation.context.ValidationContext
+import hmda.validation.dsl.{ Failure, Success }
+import hmda.validation.ValidationStats._
+import org.scalacheck.Gen
+
+import scala.concurrent.Await
+
+class Q070Spec extends MacroSpecWithValidationStats {
+  val threshold = configuration.getInt("hmda.validation.macro.Q070.currentYearThreshold")
+  val proportionSold = configuration.getDouble("hmda.validation.macro.Q070.currentYearProportion")
+  val yearDifference = configuration.getDouble("hmda.validation.macro.Q070.relativeProportion")
+
+  "Q070" must {
+    //// Check #1: comparing last year to this year ////
+    val instId = "inst-with-prev-year-data"
+    "set up: persist last year's data: sold 60% of loans" in {
+      validationStats ! AddSubmissionMacroStats(SubmissionId(instId, "2016", 1), 0, 100, 60, 0, 0, 0, 0)
+      val (relevant, relevantSold) = Await.result((validationStats ? FindQ070(instId, "2016")).mapTo[(Int, Int)], duration)
+      relevant mustBe 100
+      relevantSold mustBe 60
+    }
+    "(previous year check) pass when percentage sold is greater in current year than previous year" in {
+      val numSold = (200 * (0.6 + yearDifference)).toInt
+      val relevantNotSoldLars = listOfN(200 - numSold, Q070Spec.relevantNotSold)
+      val relevantSoldLars = listOfN(numSold, Q070Spec.relevantSold)
+      val irrelevantLars = listOfN(any, Q070Spec.irrelevant)
+      val testLars = toSource(relevantSoldLars ++ relevantNotSoldLars ++ irrelevantLars)
+      Q070.inContext(ctx(instId))(testLars).map(r => r mustBe a[Success])
+    }
+    s"(previous year check) pass when percentage sold is close enough to last year's percentage sold (within $yearDifference)" in {
+      val numSold = (200 * (0.6 - yearDifference) + 1).toInt
+      val relevantNotSoldLars = listOfN(200 - numSold, Q070Spec.relevantNotSold)
+      val relevantSoldLars = listOfN(numSold, Q070Spec.relevantSold)
+      val irrelevantLars = listOfN(any, Q070Spec.irrelevant)
+      val testLars = toSource(relevantSoldLars ++ relevantNotSoldLars ++ irrelevantLars)
+      Q070.inContext(ctx(instId))(testLars).map(r => r mustBe a[Success])
+    }
+    s"(previous year check) fail when percentage sold is too low compared to previous year ($yearDifference difference or more)" in {
+      val numSold = (200 * (0.6 - yearDifference) - 1).toInt
+      val relevantNotSoldLars = listOfN(200 - numSold, Q070Spec.relevantNotSold)
+      val relevantSoldLars = listOfN(numSold, Q070Spec.relevantSold)
+      val irrelevantLars = listOfN(any, Q070Spec.irrelevant)
+      val testLars = toSource(relevantSoldLars ++ relevantNotSoldLars ++ irrelevantLars)
+      Q070.inContext(ctx(instId))(testLars).map(r => r mustBe a[Failure])
+    }
+
+    //// Check #2: Current Year ////
+    "(current year check) fails when too few relevant loans sold to Ginnie Mae" in {
```

Comment:
This spec and the next one reference Ginnie Mae in the title (copy-paste mistake). Should be Fannie or Freddie.